scrapy/crawler.py

```diff
@@ -73,11 +74,21 @@ def crawl(self, *args, **kwargs):
             yield self.engine.open_spider(self.spider, start_requests)
             yield defer.maybeDeferred(self.engine.start)
         except Exception:
-            exc = defer.fail()
+            # In Python 2 reraising an exception after yield discards
+            # the original traceback (see http://bugs.python.org/issue7563),
+            # so sys.exc_info() workaround is used.
+            # This workaround also works in Python 3, but it is not needed,
+            # and it is slower, so in Python 3 we use native `raise`.
+            if six.PY2:
+                exc_info = sys.exc_info()
+
             self.crawling = False
             if self.engine is not None:
                 yield self.engine.close()
-            yield exc
+
+            if six.PY2:
+                raise six.reraise(*exc_info)
```

Comment:
I'm not sure, but perhaps ``raise exc_info`` will also work and will not add and extra frame of ``six.reraise`` to the traceback?